R/crunch-data-frame.R

```diff
@@ -0,0 +1,291 @@
+        } else if (!is.numeric(i)) {
+            halt("row subsetting must be done with either numeric or a logical.")
+        }
+        row_inds <- i
+    }
+    return(row_inds)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[[.CrunchDataFrame` <- function (x, i) {
+    return(getCrdfVar(grabColNames(x, i, allow_logical = FALSE),
+                      x, mode = attr(x, "mode")))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[[<-.CrunchDataFrame` <- function (x, i, value) {
+    return(setCrdfVar(col_name = grabColNames(x, i, allow_logical = FALSE),
+                      crdf = x, value = value))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`$.CrunchDataFrame` <- function (x, i) {
+    return(getCrdfVar(i, x, mode = attr(x, "mode")))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`$<-.CrunchDataFrame` <- function (x, i, value) {
+    return(setCrdfVar(col_name = i, crdf = x, value = value))
+}
+
+getCrdfVar <- function (col_name, crdf, ...) {
+    if (!col_name %in% names(crdf)) {
+        halt("The variable ", dQuote(col_name), " is not found in the CrunchDataFrame.")
+    }
+    
+    return(get(col_name, crdf)) # need [ord] here?
+}
+
+getVarFromServer <- function (col_name, crdf, ...) {
+    ord <- attr(crdf, "order")
+    ds <- attr(crdf, "crunchDataset")
+    
+    col_ref <- ds[[col_name]]
+
+    if (!is.null(ord)) {
+        return(as.vector(col_ref[ord], ...)[match(ord, sort(unique(ord)))])
+    } else {
+        # no reordering if order is null
+        return(as.vector(col_ref, ...))
+    }
+}
+
+setCrdfVar <- function (col_name, row_inds, crdf, value) {
+    # remove the variable if the value is NULL
+    if (is.null(value) & missing(row_inds)) {
+        return(invisible(rmCrdfVar(col_name = col_name, crdf = crdf)))
+    }
+    
+    if (missing(row_inds)) {
+        row_inds <- seq_len(nrow(crdf))
+    }
+    
+    # validate row lengths
+    if (length(value) == 1 & length(row_inds) > 1) {
+        # if there is a single value, but more than one row, then we need to 
+        # repeat the value for as many times as there are rows. This is the 
+        # only exception to the rule that there must be the same number of 
+        # values as rows.
+        value <- rep(value, length(row_inds))
+    } else if (is.data.frame(value)) {
+        # we got a data.frame as values, likely this is an array or MR variable
+        # Flatten it so that each sub variable is its own column, without name 
+        # prefixing. This is slightly different from what data.frame(v1 = 
+        # data.frame(a=1, b=2)) would do.
+        value <- unlist(value, recursive = FALSE)
+    } else if (length(value) != length(row_inds)) {
+        halt("replacement has ", length(value),
+             " rows, the CrunchDataFrame has ", nrow(crdf))
+    }
+    
+    # if we are replacing some rows, fill in with NAs, or grab old vector
+    if (length(row_inds) != nrow(crdf)) {
+        if (!{col_name %in% names(crdf)}) {
```

Comment:
Radford Neal may be right about { being 6% faster than (, and given his surname I'm inclined to trust him, but visually I think that using { like this is an abomination. I'll take the few extra microseconds in exchange for the elegance of the parenthesis's smooth curve.